peps/pep-0752.rst

```diff
@@ -0,0 +1,399 @@
+`corporate organizations <corp-orgs_>`_ provides the following benefits:
+
+* PyPI would have a constant source of funding for support specialists,
+  infrastructure maintenance and new features.
+* Although each application would require independent review, less human
+  feedback would be required because the process to approve a paid organization
+  already bestows a certain amount of trust.
+
+Specification
+=============
+
+Organizations MAY reserve one or more namespaces. Such reservations neither
+confer ownership nor grant special privileges to existing packages.
+
+.. _naming:
+
+Naming
+------
+
+A namespace MUST be a `valid`__ project name and `normalized`__ internally e.g.
+``foo.bar`` would become ``foo-bar``. The user facing namespace (e.g. in UI
+tooltips) MUST preserve the original casing as defined during reservation.
+
+__ https://packaging.python.org/en/latest/specifications/name-normalization/#name-format
+__ https://packaging.python.org/en/latest/specifications/name-normalization/#name-normalization
+
+Grant Semantics
+---------------
+
+A namespace grant bestows ownership over the following:
+
+1. A package matching the namespace itself such as the placeholder package
+   `microsoft <https://pypi.org/project/microsoft/>`__.
+2. Packages that start with the namespace followed by a hyphen. For example,
+   the namespace ``foo`` would match the package ``foo-bar`` but not the
+   package ``foobar``.
+
+Grant Types
+-----------
+
+There are two types of grants.
+
+.. _root-grant:
+
+Root Grant
+''''''''''
+
+Only organizations have the ability to submit requests for namespace grants.
+An organization gets a root grant for every accepted request. This grant may
+produce any number of `child grants <child-grant_>`_.
+
+.. _child-grant:
+
+Child Grant
+'''''''''''
+
+A child grant is created by the owner of a `root grant <root-grant_>`_. The
+namespace MUST be prefixed by the root grant namespace followed by a hyphen.
+For example, ``google-cloud`` would be a valid child of the root ``google``.
+
+.. _grant-ownership:
+
+Grant Ownership
+---------------
+
+The owner of a grant may allow the grant to be shared with any number of other
+organizations. Organizations with which grants are shared cannot themselves
+share such grants. For example, if organization A was given a root grant and
+decides to share this with organization B:
+
+* Organization B cannot share the grant with other organizations.
+* Organization B cannot produce a child grant.
+
+The owner may also transfer this executive ownership to another organization.
+If the organization is a corporate organization, the target for transfer must
+also be.
+
+.. _uploads:
+
+Uploads
+-------
+
+If the following criteria are all true for a given upload:
+
+1. The package does not yet exist.
+2. The name matches a reserved namespace.
+3. The user is not authorized to use the namespace by the owner of the
+   namespace.
+
+Then the upload MUST fail with a 403 HTTP status code.
```

Comment:
Is it worth clarifying somewhere that this is on a per-repository basis? I.e. if I host an internal mirror of PyPI, I can upload to whatever name I want.

Reply:
In the `Grant Semantics` section I added:

```
Namespaces are per-repository and MUST NOT be shared between repositories.
```